apidef/oas/event.go

```diff
@@ -0,0 +1,198 @@
+	return nil
+}
+
+// WebhookEvent stores the core information about a webhook event.
+type WebhookEvent struct {
+	// URL is the target URL for the webhook.
+	URL string `json:"url" bson:"url"`
+	// Method is the HTTP method for the webhook.
+	Method string `json:"method" bson:"method"`
+	// CoolDownPeriod is the cool-down for the event, so it does not trigger again (in seconds).
+	CoolDownPeriod int64 `json:"coolDownPeriod" bson:"coolDownPeriod"`
+	// BodyTemplate is the template to be used for request payload.
+	BodyTemplate string `json:"bodyTemplate,omitempty" bson:"bodyTemplate,omitempty"`
+	// Headers are the list of request headers to be used.
+	Headers Headers `json:"headers,omitempty" bson:"headers,omitempty"`
+}
+
+// GetWebhookConf converts EventHandler.WebhookEvent apidef.WebHookHandlerConf.
+func (e *EventHandler) GetWebhookConf() apidef.WebHookHandlerConf {
+	return apidef.WebHookHandlerConf{
+		Disabled:     !e.Enabled,
+		ID:           e.ID,
+		Name:         e.Name,
+		Method:       e.Webhook.Method,
+		TargetPath:   e.Webhook.URL,
+		HeaderList:   e.Webhook.Headers.Map(),
+		EventTimeout: e.Webhook.CoolDownPeriod,
+		TemplatePath: e.Webhook.BodyTemplate,
+	}
+}
+
+// EventHandlers holds the list of events to be processed for the API.
+type EventHandlers []EventHandler
+
+// Fill fills EventHandlers from classic API definition. Currently only webhook events are supported.
+func (e *EventHandlers) Fill(api apidef.APIDefinition) {
+	if len(api.EventHandlers.Events) == 0 {
+		return
+	}
+
+	events := EventHandlers{}
+	for gwEvent, ehs := range api.EventHandlers.Events {
+		for _, eh := range ehs {
+			if eh.Handler != event.WebHookHandler {
+				continue
+			}
+
+			whConf := apidef.WebHookHandlerConf{}
+			err := whConf.Scan(eh.HandlerMeta)
+			if err != nil {
+				continue
+			}
+
+			ev := EventHandler{
+				Enabled: !whConf.Disabled,
+				Trigger: gwEvent,
+				Kind:    WebhookKind,
+				ID:      whConf.ID,
+				Name:    whConf.Name,
+				Webhook: WebhookEvent{
+
+					URL:            whConf.TargetPath,
+					Method:         whConf.Method,
+					Headers:        NewHeaders(whConf.HeaderList),
+					CoolDownPeriod: whConf.EventTimeout,
+					BodyTemplate:   whConf.TemplatePath,
+				},
+			}
+
+			events = append(events, ev)
+		}
+	}
+
+	*e = events
+}
+
+// ExtractTo extracts events to apidef.APIDefinition.
+func (e *EventHandlers) ExtractTo(api *apidef.APIDefinition) {
+	if e == nil || len(*e) == 0 {
+		return
+	}
+
+	if api.EventHandlers.Events == nil {
+		api.EventHandlers.Events = make(map[apidef.TykEvent][]apidef.EventHandlerTriggerConfig)
+	}
+
+	// this blocks helps with extracting OAS into APIDefinition.
+	// update this when new event handlers are added to OAS support.
+	for eventType, eventTriggers := range api.EventHandlers.Events {
+		triggersExcludingWebhooks := make([]apidef.EventHandlerTriggerConfig, 0)
+		for _, eventTrigger := range eventTriggers {
+			if eventTrigger.Handler == event.WebHookHandler {
```

Comment:
make it a switch, `exhaustive` golangci-lint rules will report unhandled extractions